source/Font.cpp

```diff
@@ -344,143 +359,177 @@ void Font::ShowUnderlines(bool show)
-	
-	GLfloat vertices[] = {
-		   0.f,    0.f, 0.f, 0.f,
-		   0.f, glyphH, 0.f, 1.f,
-		glyphW,    0.f, 1.f, 0.f,
-		glyphW, glyphH, 1.f, 1.f
-	};
-	glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
-	
-	// connect the xy to the "vert" attribute of the vertex shader
-	glEnableVertexAttribArray(shader.Attrib("vert"));
-	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, 4 * sizeof(GLfloat), nullptr);
-	
-	glEnableVertexAttribArray(shader.Attrib("corner"));
-	glVertexAttribPointer(shader.Attrib("corner"), 2, GL_FLOAT, GL_FALSE,
-		4 * sizeof(GLfloat), (const GLvoid*)(2 * sizeof(GLfloat)));
+	for(++pos; pos < str.length(); ++pos)
+		if((str[pos] & 0x80) == 0 || (str[pos] & 0xc0) == 0xc0)
+			break;
+	return pos;
+}
+
+
+
+// Returns the start of the unicode code point at pos in utf8.
+size_t Font::CodePointStart(const string &str, size_t pos)
+{
+	// 0xxxxxxx and 11?????? start a code point
+	while(pos > 0 && (str[pos] & 0x80) != 0x00 && (str[pos] & 0xc0) != 0xc0)
+		--pos;
+	return pos;
+}
+
+
+
+// Decodes a unicode code point in utf8.
+// Invalid codepoints are converted to 0xFFFFFFFF.
+char32_t Font::DecodeCodePoint(const string &str, size_t pos)
+{
+	if(pos >= str.length())
+		return 0;
 	
-	glBindBuffer(GL_ARRAY_BUFFER, 0);
-	glBindVertexArray(0);
+	// invalid (-1) or end (0)
+	int bytes = CodePointBytes(str.c_str() + pos);
+	if(bytes < 1)
+		return bytes;
 	
-	// We must update the screen size next time we draw.
-	screenWidth = 0;
-	screenHeight = 0;
+	// 1 byte
+	if(bytes == 1)
+		return (str[pos] & 0x7f);
 	
-	// The texture always comes from texture unit 0.
-	glUniform1i(shader.Uniform("tex"), 0);
-
-	colorI = shader.Uniform("color");
-	scaleI = shader.Uniform("scale");
-	glyphI = shader.Uniform("glyph");
-	aspectI = shader.Uniform("aspect");
-	positionI = shader.Uniform("position");
+	// 2-4 bytes
+	char32_t c = (str[pos] & ((1 << (7 - bytes)) - 1));
+	for(int i = 1; i < bytes; ++i)
+		c = (c << 6) + (str[pos + i] & 0x3f);
+	return c;
+}
+
+
+
+// Prepare a string for processing by multiple sources, producing source-end pairs.
+vector<pair<size_t,size_t>> Font::Prepare(const std::string &str) const
+{
+	// XXX This is an experimental attempt at using different glyph sources to get full unicode coverage:
+	//  - assumes code points supported by multiple sources are treated the same way
+	//  - prefers the first source that supports [start,end[
+	//  - uses source 0 for unsupported data
+	vector<pair<size_t,size_t>> sections;
+	for(size_t start = 0; start < str.length(); )
+	{
+		bool isUnsupported = true;
+		
+		// Supported data.
+		for(size_t i = 0; i < sources.size(); ++i)
+		{
+			size_t end = sources[i]->FindUnsupported(str, start);
+			if(end == start)
+				continue;
+			if(!sections.empty() && sections.back().first == i)
+				sections.back().second = end;
+			else
+				sections.emplace_back(i, end);
+			isUnsupported = false;
+			start = end;
+			break;
+		}
+		
```

Comment:
This longest-match logic can choose a glyph of a lower priority font.

For instance, when I have a font setting like this
```
font
	size 14
	atlas "images/font/ubuntu14r.png"
	freetype "images/fonts/ipagp.ttf"
```
and let it render a string "あA", then "A" is rendered by ipagp.ttf because ipagp.ttf contains both glyphs of "あ" and "A."

I expect that "A" is rendered by ubuntu14r.png.

Reply:
I guess that is a preference behavior.. to me it seems sensible to use the same font to render the characters in a given string if possible, and only do piecemeal when that cannot be done.

However I can see that it would be weird if 99/100 sentences used an Ubuntu "A" and then the one sentence with a glyph has a different look for every other character too.

Reply:
Before implementing it like that I debated myself over this particular issue and came to the conclusion that both cases are perfectly valid preferences for various circumstances.

Since the objective is to get full unicode support (which needs multiple font files), in the long term the most likely case is to have multiple fonts that are compatible with each other and cover all the ranges. The noto fonts from google are aiming for this.

Another way to overcome this is to have styling that allows font selection.